components/VFBMain.js

```diff
@@ -0,0 +1,1225 @@
+                        compositeInstances.push(visualInstances[i]);
+                    }
+                }
+                return compositeInstances;
+            });
+        }
+
+        // Query builder initialization
+        this.refs.querybuilderRef.setResultsColumnMeta(this.queryResultsColMeta);
+        this.refs.querybuilderRef.setResultsColumns(this.queryResultsColumns);
+        this.refs.querybuilderRef.setResultsControlsConfig(this.queryResultsControlConfig);
+        this.refs.querybuilderRef.addDataSource(this.queryBuilderDatasourceConfig);
+
+        // Loading ids passed through the browser's url
+        var idsList = this.props.location.search.replace("?i=", "");
+        if((idsList.length > 0) && (this.state.modelLoaded == true) && (this.urlIdsLoaded == false)) {
+            this.urlIdsLoaded = true;
+            var idArray = idsList.split(",");
+            var that = this;
+            console.log("Loading IDS from url");
+            GEPPETTO.on(GEPPETTO.Events.Model_loaded, function () {
+                that.addVfbId(idArray);
+            });
+        }
+
+        // Selection listener
+        GEPPETTO.on(GEPPETTO.Events.Select, function (instance) {
+            var selection = GEPPETTO.SceneController.getSelection();
+            if (selection.length > 0 && instance.isSelected()) {
+                var latestSelection = instance;
+                var currentSelectionName = "";
+                if (this.refs.termInfoRef.getTermInfoWidget() != undefined) {
+                    currentSelectionName = this.refs.termInfoRef.getTermInfoWidget().name;
+                }
+                if (latestSelection.getChildren().length > 0) {
+                    // it's a wrapper object - if name is different from current selection set term info
+                    if (currentSelectionName != latestSelection.getName()) {
+                        this.refs.termInfoRef.setTermInfo(latestSelection[latestSelection.getId() + "_meta"], latestSelection[latestSelection.getId() + "_meta"].getName());
+                    }
+                } else {
+                    // it's a leaf (no children) / grab parent if name is different from current selection set term info
+                    var parent = latestSelection.getParent();
+                    if (parent != null && currentSelectionName != parent.getName()) {
+                        this.refs.termInfoRef.setTermInfo(parent[parent.getId() + "_meta"], parent[parent.getId() + "_meta"].getName());
+                    }
+                }
+            }
+            if (window.StackViewer1 != undefined) {
+                this.refs.stackViewerRef.updateStackWidget();
+            }
+        }.bind(this));
+    };
+
+    // Children handlers
+    buttonBarHandler(buttonState) {
+        if((buttonState !== "infoBtn") || (buttonState !== "stackBtn")) {
+            var tempState = this.state[buttonState];
+            this.setState({
+                [buttonState] : !tempState
```

Comment:
I see what you are trying to do here saving some code and why the states are called what they are, the downside of this though is that the logic is less readable. For anybody who hasn't written the code they will look at those states and won't know what they do just by reading their name. 

infoBtn: true vs termInfoVisible: true
stackBtn: true vs stackViewerVisible: true
tutorialBtn: false vs tutorialVisible: false
etc.

One way you could do this is renaming the ids in the buttonBar configuration to be just "tutorial" and inside this function append the string "visible" before accessing the state. In general it's a fine line with these "smart" tricks before we write unnecessarily hard to read code so better to always be careful and err on the side of simplicity at the cost of more lines.

Reply:
done 👍